provider/pkg/schema/jsschema_util.go

```diff
@@ -0,0 +1,203 @@
+// Copyright 2016-2021, Pulumi Corporation.
+
+package schema
+
+import (
+	jsschema "github.com/lestrrat-go/jsschema"
+)
+
+// FlattenJSSchema recursively flattens a schema containing AnyOf or OneOf into a list of schemas
+// Note: this only performs a shallow copy, don't use in situations where schema fields may be mutated.
+// TODO: this should probably handle AllOf and Not as well.
+func FlattenJSSchema(sch *jsschema.Schema) jsschema.SchemaList {
```

Comment:
Should we add tests for these new functions?

Reply:
These are a little bit of a hassle to test; I've tested the cases we care about most in gen_examples_test, but I'll add some property tests for these two to boost coverage.